[OptimistLabyrinth] 🚀 3단계 - 질문 삭제하기 리팩터링

OptimistLabyrinth opened the pull request.

Commit:
chore: 불필요한 import 삭제

Commit:
chore: 불변성 멤버변수에 final 키워드 추가

Commit:
feat: 복잡한 서비스 로직을 단위테스트 가능하도록 도메인 레이어에 구현

도메인 레이어에 대한 단위 테스트 작성 후
서비스 레이어 코드 리팩터링 예정

Commit:
test: 테스트 헬퍼 메서드 공용으로 사용하기 위해 별도의 클래스로 분리

QnaServiceTest 뿐만 아니라 다른 테스트 코드에서도 사용할 수 있도록
별도 클래스로 분리

Commit:
test: 방금 추가한 Question.delete, Answer.delete 메서드에 대한 테스트 코드 추가

Commit:
feat: QnaService.deleteQuestion 메서드 리팩터링

동일한 사용자인지 확인할 때 User 객체 자체를 비교하는 대신에 id 값으로 비교
기존의 QnaServiceTest 의 모든 테스트 케이스 그대로 성공

mskangg commented on src/main/java/qna/domain/Question.java:
```diff
@@ -61,6 +64,33 @@ public Question setDeleted(boolean deleted) {
+        return deleteHistories;
+    }
+
+    @Getter
+    @AllArgsConstructor
+    public static class DeleteResultDto {
+        private final Question question;
+        private final List<Answer> answers;
+        private final List<DeleteHistory> deleteHistories;
+    }
```
domain 객체에 dto를 추가하셨네요.
내부 상태로 갖고있는 answers와는 무엇이 다른가요?? 🤔

OptimistLabyrinth replied:
리턴값으로 여러개의 객체를 돌려주고 싶어서 DTO 객체를 만들었어요.
애초에 리턴값이 여러개가 되는 것 자체가 잘못된 설계인걸까요??

mskangg commented on src/main/java/qna/domain/Question.java:
```diff
@@ -61,6 +64,33 @@ public Question setDeleted(boolean deleted) {
+        }
+        for (final Answer answer : answers) {
+            answer.delete(loginUser);
+        }
```
질문 삭제시 연관된 답변들을 삭제하는 로직 중 `answer.delete(loginUser)` 에서 리턴값을 `DeleteHistory`로 구현한다면  `DeleteResultDto`은 필요하지 않아보입니다.

이런식으로도 구현해볼 수 있을 것 같아요!
이미 지운 답변들의 히스토리를 생성해서 넣는 `getDeleteHistories()` 메서드도 리팩토링 할 수 있어보입니다.

```java
    public List<DeleteHistory> delete2(User loginUser) {
        validateOwner(loginUser);

        List<DeleteHistory> deleteHistories = this.answers.stream()
                .map(answer -> answer.delete2(loginUser))
                .collect(toList());

        setDeleted(true);

        deleteHistories.add(new DeleteHistory(id, ContentType.QUESTION, loginUser));
        return deleteHistories;
    }
```

OptimistLabyrinth replied:
수정 완료했습니다.

mskangg commented on src/main/java/qna/domain/Question.java:
```diff
@@ -61,6 +64,33 @@ public Question setDeleted(boolean deleted) {
         return this;
     }
 
+    public DeleteResultDto delete(User loginUser) throws CannotDeleteException {
```
CanNotDeleteException이 현재 checkedException으로 설정되어있는데, uncheckedException으로 변경해도 좋을 것 같아요 :)
변경하지 않는다면 예외를 전파하는 것이 아닌 check 해서 예외의 책임을 다하게 구현하시는게 좋아보입니다.

OptimistLabyrinth replied:
수정했습니다.

mskangg commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -34,27 +30,10 @@ public Question findQuestionById(Long id) {
+        final Question.DeleteResultDto deleteResultDto = question.delete(loginUser);
         questionRepository.save(question);
-        answerRepository.saveAll(answers);
-        deleteHistoryService.saveAll(deleteHistories);
+        answerRepository.saveAll(deleteResultDto.getAnswers());
```
question과 answers를 다시 save 하신 이유가 있으실까요??

OptimistLabyrinth replied:
제가 JPA 에서 영속성 콘텍스트가 어떻게 운영되는지 확실하게 몰라서
수정한 모든 데이터에 대해서 save 호출했어요.
하나의 엔티티에 대해서만 save 호출하면 연관성 있는 모든 엔티티도 함께 저장이 되나요??

다른 언어에서는 모든 엔티티에 대해서 명시적으로 save 호출해야만 제대로 동작하는 경우들이 있거든요

OptimistLabyrinth replied:
redundant XXXRepository.save( ) 메서드 호출 제거 완료했습니다 ㅎㅎ

mskangg reviewed (changes requested):
3단계 미션 잘 구현해주셨습니다.
다만, 테스트코드가 동작하지 않는 것들이 있습니다.

<img width="311" alt="image" src="https://user-images.githubusercontent.com/27044096/204101669-b383b6ec-57a9-47bc-98ca-a1492c1a11f6.png">

테스트는 독립적인 환경에서 돌아가야하니 꼭 확인 부탁드려요!

그리고 몇가지 피드백 드렸으니 확인 부탁드립니다.

Commit:
fix: 어떤 머신에서 실행하든 H2 데이터베이스로 실행하도록 수정

Commit:
feat: 피드백 내용 반영

[1] checked exception 대신에 unchecked exception 사용하기
[2] Answer.delete, Question.delete 메서드에서 DeleteHistory 반환하도록 수정
[3] 불필요한 XXXRepository.save 호출 제거

OptimistLabyrinth commented:
application.properties 파일이 문제였을까요??

spring.datasource.url 만 세팅하는 경우에 통합테스트에서 오류가 발생하나요??

- spring.datasource.driverClassName
- spring.jpa.database-platform

이 두 개의 값도 세팅한 뒤에 다시 PR 요청드려봅니다.

![Screenshot 2022-11-27 at 11 18 00 AM](https://user-images.githubusercontent.com/16411622/204116276-49be6b06-db6a-46bd-a2dc-e3c87f8bd3a9.png)

mskangg reviewed (approved):
피드백반영 잘해주셨어요! 👍 
모든 미션 완료하셨습니다,,!
과정의 절반이 지나고 있는 시점이니 시작보단 지치셨을 것 같아요!
페이스 조절 잘하셔서 꼭 마지막 미션까지 완료하셨으면 좋겠습니다!! 🙇‍♂️
이번 JPA 과정 정말정말 고생하셨고 남은 미션들도 화이팅입니다! 🔥
